platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,447 @@
+We propose the introduction of a queue of roster hashes stored in the state, where the index of a roster hash determines
+what round the roster is active for as shown.
+![](TSS%20Roster%20Lifecycle-Roster%20and%20Rounds.drawio.svg)
+The size of this queue would be ancient round window + roster offset.
+The roster offset refers to the number of rounds in advance that the network determines and declares the active roster
+for a specific future round.
+This acts as a proactive mechanism that allows the network to prepare for upcoming changes in the roster and ensure a
+valid transition to a future roster.
+When each round is handled, we will pop off the active roster hash for the round that just became ancient, and add a new
+active roster hash for a new future round.
+
+Before full DAB (DAB Phase 2.5), we propose to store the current and previous active rosters in this newly introduced
+queue of hashes.
+In the DAB Phase 2.5 implementation, this queue will contain only 2 hashes, with the first hash representing
+the previous active roster, and the second hash representing the current active roster.
+At upgrade boundaries, we will pop off the previous active roster, and add a new roster to the end of the queue, which
+is the new active roster.
+
+This approach provides the benefit of introducing the necessary data structure for full DAB, while also providing a
+mechanism for tracking the active roster for each round in the interim.
+
+#### Reconnect
+
+Reconnect. Reconnect logic currently exchanges and validates the address book between the learner and teacher nodes. The
+  learner node uses the address book to select the teacher node, as well as remove all invalid signatures from the
+  state. Both of these will need to be updated to use rosters instead.
+
+#### Others
+- Networking. Most parts of the network code have been abstracted away from the address book as part of the Dynamic
+  Address Book effort, so there should be minimal work left there.
+- Event Validation. The event validation logic uses the address book to determine whether a given event has a valid
+  signature. This will be updated to use rosters instead.
+- Miscellaneous components. Other components that reference things stored in the address book (like consensus needing
+  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.
+
+### Test Plan
+
+Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:
+
+1. New valid Candidate Roster created with no subsequent one sent by app. Verify accept.
+2. New valid Candidate Roster created with subsequent one sent by app. Verify accept.
+3. Invalid roster(s) sent by the app. Verify reject.
+4. Node Failures During Roster Change: What happens if nodes fails or disconnects during a roster change? Verify valid
+   node successfully reconnects.
```

Comment:
Actually, this one is interesting. Suppose a node is dead during a software upgrade when the rest of the network adopts a new roster.

Once the dead node resumes after the network has adopted a new roster, will it be able to rejoin the network? At this point, this node has a pointer to the network's current active roster in its candidate roster pointer.

To complicate this a bit further, suppose the node has been dead during two consecutive software upgrades and the network has replaced the roster during both, so the dead node has absolutely no references to the current roster at all, even as a candidate roster. And the rest of the network has already forgotten about the "previous roster" that was used when this node was last alive. Will it still be able to reconnect?

Do we even support such scenarios or we require all nodes to be alive during every software upgrade/roster adoption event? If so, I think we should document this assumption to ensure we don't miss any edge cases that we need to support.

Reply:
The node will be able to rejoin the network because it's key is in the active roster. If it's able to join the network and reconnect, it will get the latest state which will contain the active roster. @edward-swirdlabs, is that correct?

Reply:
Are nodes allowed to reconnect with each other across software upgrade boundaries?   Or do nodes have to be on the same software version to perform a reconnect?  

Suppose a node is down for a long time and doesn't receive the freeze-command.  The rest of the network performs a software upgrade.  The node is restarted and comes back up.  What happens?  @litt3 may know off the top of his head.  Otherwise we'll need to look at the code. 

I don't expect anything to change about reconnect with the TSS-Roster proposal.    We might need to speak with services about what would happen in response to DAB Phase 2.  Our behavior will be inherited from where they leave off.

Reply:
If I understand correctly, a node operator would have to upgrade the software (apparently, upload latest .jars or similar) to the node.

Assuming this is correct, and assuming that the node still has a state on disk saved with the previous version of the software, or, in the second scenario, with two-versions-old software, would it be able to rejoin the network then?

The question is more about whether we support such scenarios officially rather than whether the current code is capable of doing this. If we do support them, then we need tests for these scenarios. If we don't support them, then let's explicitly state so in the testing plan.

Reply:
@anthony-swirldslabs - i think this is a valid question. Mind asking in one of the channels?

Reply:
Just posted a question to the engineering-team channel.

Reply:
@kfa-aguda - so, you can find all the details in that slack channel. But long story short, both the scenarios above are supported.

To summarize and rephrase, a node X can be dead for as long as it pleases, even for years, hence skipping multiple roster updates. It should be able to rejoin the network w/o any manual interventions as long as both the following conditions are true:
1. The node X itself remains in the roster during all this time. In fact, I think it may be removed in the middle of that period, but the main point is that it was present in the roster when it died, and it is present in the roster when it comes back alive.
2. The old version of the roster on node X that was preserved when it died should contain at least one node (e.g. node Y) that is still active on the network when the node X comes back alive. It is so that the node X can connect to at least that node Y in order to start a reconnect session and catch up with the network.

While the above is mostly handled by the reconnect itself, the Roster proposal introduces a new implication because it preserves the references to the active and candidate (and potentially, one or many past rosters) in the state. So we need to test at least two scenarios explicitly as described in the original comment:
1. The node X skips a single upgrade and goes through it after the rest of the network has upgraded and updated the roster.
2. The node X skips several upgrades/several roster changes, and comes back online at a later time. As long as the conditions described above are satisfied, the node should still be able to rejoin and update its roster to the latest one.

Please include these test scenarios in the testing plan.

Reply:
Thanks for bringing the update here @anthony-swirldslabs. Just saw the conversation. I think (for completeness purposes) there is one more condition discussed - that the node X runs the same software version as the rest of the network. I'll add these test cases as described.